### PR TITLE
Block sample when honesty is on

### PR DESCRIPTION
**Pull request by MathLens**

### CODE REVIEW

**Comment by MathLens** [comment]
@chenhongqi0716 you can try to implement the function.

**Comment by chenhongqi0716** [comment]
> @chenhongqi0716 you can try to implement the function.

I finished. You can have a look and check, maybe there is an easy way to do this. BTW, how can I test this function?

**Comment by MathLens** [comment]
> > @chenhongqi0716 you can try to implement the function.
> 
> I finished. You can have a look and check, maybe there is an easy way to do this. BTW, how can I test this function?

You can either copy it to an empty project to test, or you create a test executable in this repo. The latter will be a little involved, so maybe next time.

**Comment by chenhongqi0716** [comment]
> > > @chenhongqi0716 you can try to implement the function.
> > 
> > 
> > I finished. You can have a look and check, maybe there is an easy way to do this. BTW, how can I test this function?
> 
> You can either copy it to an empty project to test, or you create a test executable in this repo. The latter will be a little involved, so maybe next time.

Great. It seems there is no problem. I feel like I return the first year in sjtu.

**Comment by MathLens** [comment]
@chenhongqi0716 
Now we also need the block length in TreeOptions (as what we did for SamplingOptions): https://github.com/chenhongqi0716/grf/blob/honesty/core/src/tree/TreeOptions.h#L38

Can you add a data member in this class and change the interface accordingly so that it also gets block length when mbb is true?

**Comment by chenhongqi0716** [comment]
> @chenhongqi0716
> Now we also need the block length in TreeOptions (as what we did for SamplingOptions): https://github.com/chenhongqi0716/grf/blob/honesty/core/src/tree/TreeOptions.h#L38
> 
> Can you add a data member in this class and change the interface accordingly so that it also gets block length when mbb is true?

I guess I should revise it as above.

**Comment by chenhongqi0716** [comment]
One more question. In the SamplingOptions, should we also set the default block_length as 1?

https://github.com/chenhongqi0716/grf/blob/4c9be457432fece9444bcb8fa042da3d467e5f6b/core/src/sampling/SamplingOptions.h#L59

**Comment by MathLens** [comment]
> One more question. In the SamplingOptions, should we also set the default block_length as 1?
> 
> https://github.com/chenhongqi0716/grf/blob/4c9be457432fece9444bcb8fa042da3d467e5f6b/core/src/sampling/SamplingOptions.h#L59

Yes. Will do that later when modifying other functions in the samplers.

**Comment by MathLens** [comment]
> > @chenhongqi0716
> > Now we also need the block length in TreeOptions (as what we did for SamplingOptions): https://github.com/chenhongqi0716/grf/blob/honesty/core/src/tree/TreeOptions.h#L38
> > Can you add a data member in this class and change the interface accordingly so that it also gets block length when mbb is true?
> 
> I guess I should revise it as above.

Also need to change the code where the constructor getting called.

**Comment by chenhongqi0716** [comment]
> > > @chenhongqi0716
> > > Now we also need the block length in TreeOptions (as what we did for SamplingOptions): https://github.com/chenhongqi0716/grf/blob/honesty/core/src/tree/TreeOptions.h#L38
> > > Can you add a data member in this class and change the interface accordingly so that it also gets block length when mbb is true?
> > 
> > 
> > I guess I should revise it as above.
> 
> Also need to change the code where the constructor getting called.

I suppose we should also revise this. Does it mean that we need to revise the constructor of ForestOptions?

https://github.com/chenhongqi0716/grf/blob/7c32454c60b0819a7ae4548082484c3a5532991a/core/src/forest/ForestOptions.cpp#L43

**Comment by MathLens** [comment]
> > > > @chenhongqi0716
> > > > Now we also need the block length in TreeOptions (as what we did for SamplingOptions): https://github.com/chenhongqi0716/grf/blob/honesty/core/src/tree/TreeOptions.h#L38
> > > > Can you add a data member in this class and change the interface accordingly so that it also gets block length when mbb is true?
> > > 
> > > 
> > > I guess I should revise it as above.
> > 
> > 
> > Also need to change the code where the constructor getting called.
> 
> I suppose we should also revise this. Does it mean that we need to revise the constructor of ForestOptions?
> 
> https://github.com/chenhongqi0716/grf/blob/7c32454c60b0819a7ae4548082484c3a5532991a/core/src/forest/ForestOptions.cpp#L43

yes, I think so. We already did last time.

**Comment by chenhongqi0716** [comment]
> > > > > @chenhongqi0716
> > > > > Now we also need the block length in TreeOptions (as what we did for SamplingOptions): https://github.com/chenhongqi0716/grf/blob/honesty/core/src/tree/TreeOptions.h#L38
> > > > > Can you add a data member in this class and change the interface accordingly so that it also gets block length when mbb is true?
> > > > 
> > > > 
> > > > I guess I should revise it as above.
> > > 
> > > 
> > > Also need to change the code where the constructor getting called.
> > 
> > 
> > I suppose we should also revise this. Does it mean that we need to revise the constructor of ForestOptions?
> > https://github.com/chenhongqi0716/grf/blob/7c32454c60b0819a7ae4548082484c3a5532991a/core/src/forest/ForestOptions.cpp#L43
> 
> yes, I think so. We already did last time.

Can I just write it as above?

**Comment by chenhongqi0716** [comment]
Do we need to rewrite the constructor in TreeOptions?

File: core/src/tree/TreeOptions.cpp [hunk 1]
```diff
@@ -25,14 +25,16 @@ TreeOptions::TreeOptions(uint mtry,
                          double honesty_fraction,
                          bool honesty_prune_leaves,
                          double alpha,
-                         double imbalance_penalty):
+                         double imbalance_penalty,
+                         uint block_length):
```
Comment (MathLens) [inline core/src/tree/TreeOptions.cpp:29]: need a default argument here, in case mbb=false

Comment (chenhongqi0716) [reply core/src/tree/TreeOptions.cpp:29]: done

**Comment by MathLens** [comment]
> Do we need to rewrite the constructor in TreeOptions?

What do you mean by rewriting?

**Comment by chenhongqi0716** [comment]
> > Do we need to rewrite the constructor in TreeOptions?
> 
> What do you mean by rewriting?

I thought we needed to add another constructor at the beginning to deal with the cases mbb=T/F separately. Now we can just set the default value.

**Comment by MathLens** [comment]
Oh, you mean function overloading. An alternative to function overloading is using default values, which is simpler but less flexible.

**Review by chenhongqi0716 — COMMENTED** [review]
If we have this https://github.com/chenhongqi0716/grf/blob/ddc9a77979cf49af8628268b6aaeaed1c445b5ae/core/src/tree/TreeOptions.h#L78

Do we still need to add the default value here?

**Comment by MathLens** [comment]
> If we have this
> 
> https://github.com/chenhongqi0716/grf/blob/ddc9a77979cf49af8628268b6aaeaed1c445b5ae/core/src/tree/TreeOptions.h#L78
> 
> Do we still need to add the default value here?

Sorry. I pointed to the wrong line. The default value should always be in the declaration in the .h, not the .cpp. And this default value is unnecessary if we already set it in constructor.

**Comment by chenhongqi0716** [comment]
> > If we have this
> > https://github.com/chenhongqi0716/grf/blob/ddc9a77979cf49af8628268b6aaeaed1c445b5ae/core/src/tree/TreeOptions.h#L78
> > 
> > Do we still need to add the default value here?
> 
> Sorry. I pointed to the wrong line. The default value should always be in the declaration in the .h, not the .cpp. And this default value is unnecessary if we already set it in constructor.

Gotcha, I removed the previous update.

**Comment by MathLens** [comment]
> > > If we have this
> > > https://github.com/chenhongqi0716/grf/blob/ddc9a77979cf49af8628268b6aaeaed1c445b5ae/core/src/tree/TreeOptions.h#L78
> > > 
> > > Do we still need to add the default value here?
> > 
> > 
> > Sorry. I pointed to the wrong line. The default value should always be in the declaration in the .h, not the .cpp. And this default value is unnecessary if we already set it in constructor.
> 
> Gotcha, I removed the previous update.

and need to add it to .h instead

**Comment by chenhongqi0716** [comment]
> > > > If we have this
> > > > https://github.com/chenhongqi0716/grf/blob/ddc9a77979cf49af8628268b6aaeaed1c445b5ae/core/src/tree/TreeOptions.h#L78
> > > > 
> > > > Do we still need to add the default value here?
> > > 
> > > 
> > > Sorry. I pointed to the wrong line. The default value should always be in the declaration in the .h, not the .cpp. And this default value is unnecessary if we already set it in constructor.
> > 
> > 
> > Gotcha, I removed the previous update.
> 
> and need to add it to .h instead

It already exists. What do you mean by adding again?

**Comment by MathLens** [comment]
> > > > > If we have this
> > > > > https://github.com/chenhongqi0716/grf/blob/ddc9a77979cf49af8628268b6aaeaed1c445b5ae/core/src/tree/TreeOptions.h#L78
> > > > > 
> > > > > Do we still need to add the default value here?
> > > > 
> > > > 
> > > > Sorry. I pointed to the wrong line. The default value should always be in the declaration in the .h, not the .cpp. And this default value is unnecessary if we already set it in constructor.
> > > 
> > > 
> > > Gotcha, I removed the previous update.
> > 
> > 
> > and need to add it to .h instead
> 
> It already exists. What do you mean by adding again?

I meant this assignment is unnecessary if we have set the value in the constructor. We were talking about function overloading/default arguments of the constructor.

https://github.com/chenhongqi0716/grf/blob/ddc9a77979cf49af8628268b6aaeaed1c445b5ae/core/src/tree/TreeOptions.h#L78

We need to set the value in the declaration of the constructor instead of the corresponding line in the .cpp.

**Comment by chenhongqi0716** [comment]
> 
> I meant this assignment is unnecessary if we have set the value in the constructor. We were talking about function overloading/default arguments of the constructor.
> 
> https://github.com/chenhongqi0716/grf/blob/ddc9a77979cf49af8628268b6aaeaed1c445b5ae/core/src/tree/TreeOptions.h#L78
> 
> We need to set the value in the declaration of the constructor instead of the corresponding line in the .cpp.

Oh, I see. Thanks. It should be ok now.

**Review by chenhongqi0716 — COMMENTED** [review]
something is not right here. While compiling, the error info is: 


src/forest/ForestOptions.cpp:85:125:error:  match for call to ‘(grf::TreeOptions) (grf::uint&, grf::uint&, bool&, double&, bool&, double&, double&, grf::uint&)
      85 |     tree_options(mtry, min_node_size, honesty, honesty_fraction, honesty_prune_leaves, alpha, imbalance_penalty, blocklength)
         |                                                                                                                             ^
   make: *** [/usr/lib/R/etc/Makeconf:177: src/forest/ForestOptions.o] Error 1
   ERROR: compilation failed for package ‘grf’

Do you know the problem? Thanks!

File: core/src/forest/ForestOptions.cpp [hunk 2]
```diff
@@ -82,6 +82,7 @@ ForestOptions::ForestOptions(uint num_trees,
     ForestOptions(num_trees, ci_group_size, sample_fraction, mtry, min_node_size,
               honesty, honesty_fraction, honesty_prune_leaves, alpha, imbalance_penalty,
               num_threads, random_seed, sample_clusters, samples_per_cluster){
+    tree_options(mtry, min_node_size, honesty, honesty_fraction, honesty_prune_leaves, alpha, imbalance_penalty, blocklength);
```
Comment (MathLens) [inline core/src/forest/ForestOptions.cpp]: @chenhongqi0716 This does not work. Cannot initialize a data member here. I fixed it.

**Comment by MathLens** [comment]
I think you start testing

**Comment by chenhongqi0716** [comment]
> I think you start testing

Something appears not right for the original iid setup. e.g. "regression_forest(X,Y, num.trees = 2000, mbb = F)". Let me try to figure it out.

**Comment by MathLens** [comment]
> > I think you start testing
> 
> Something appears not right for the original iid setup. e.g. "regression_forest(X,Y, num.trees = 2000, mbb = F)". Let me try to figure it out.

My bad. Hope it is fixed now.

**Comment by chenhongqi0716** [comment]
> > > I think you start testing
> > 
> > 
> > Something appears not right for the original iid setup. e.g. "regression_forest(X,Y, num.trees = 2000, mbb = F)". Let me try to figure it out.
> 
> My bad. Hope it is fixed

Great! now it works.

**Comment by chenhongqi0716** [comment]
Can we merge it to dev?

**Review by MathLens — COMMENTED** [review]
Looks good to me

**Pull request merged**